### DISCUSSION TITLE
Shared subscription with persistence

### DISCUSSION THREAD

**Discussion by obschepit** [emqx/emqx#4576] [Q&A]
Hi.
I downloaded the distribution EMQX Enterprise v.4.2.5 Debian 10 to see persistence.
I'm interested persistence for shared subscriptions.
I found strange behavior for this case.

Preparation:
- there are 2 subscriptions (clientA and clientB) with QOS = 1 and Clean Session = False,
- one publisher with QOS = 1 (clientC),
- broker_shared_subscription_strategy = sticky (for clarity. but it does not matter).

Step 1:
1. clientA subscribes to topic $share/group1/testtopic/#
2. clientB subscribes to topic $share/group1/testtopic/#
3. clientC publishes three messages to topic testtopic/111

= clientB receives three messages - OK

Step 2:
1. clientB was disconnected
2. clientC publishes three messages to topic testtopic/111

= clientA receives three messages - OK

Step 3:
1. clientA was disconnected
2. clientC publishes three messages to topic testtopic/111
3. clientA connected
4. clientB connected

= clientA receive one message and clientB receive two messages - NOT OK

In my opinion, there should be a behavior in which the first connected should receive all messages. And this is regardless of the value of broker_shared_subscription_strategy.

What do you think about this?

**Comment by zmstone** [comment]
Hi @obschepit
How long was it between the events in step 3 ?
and did client reconnect ?

**Reply by obschepit** [reply]
seconds 5 maybe.
yes, clients reconnected. - clientA receive one message and clientB receive two messages.

**Reply by zmstone** [reply]
Sorry, I meant to ask: did the producer client reconnect

**Reply by zmstone** [reply]
If there were no subscribers connected before event 2, and event 3 was five seconds later, I wonder why should any of the subscribers receive any of the three messages.

**Reply by obschepit** [reply]
producer (clientC) did not reconnect. steps follow each other.

**Comment by zmstone** [comment]
Will try to reproduce it

**Reply by obschepit** [reply]
thank you!

**Comment by zmstone** [comment]
Hi @obschepit 
I can't seem to reproduce it.

Could you double check if the strategy is `sticky` by:
1. `emqx/bin/emqx attach` to attach to the Erlang console
2.  evaluate Erlang expression: `application:get_env(emqx, shared_subscription_strategy).`

Should expect a `{ok,sticky}` printed in Erlang console.

my test steps for ref:
One publisher: `mosquitto_pub -h localhost -t 'testtopic/111' -q 1 -l`
I use the `-l` publisher option to keep the connection open (but not to reconnect each command).

Two subscribers: `mosquitto_sub -h localhost -t '$share/group1/testtopic/#'`

**Reply by obschepit** [reply]
Hi!
1. Yes. a {ok,sticky} printed in Erlang console. But as I wrote - strategy doesn't matter.
2. I use --disable-clean-session on my mosquitto_sub. The problem is persistence.

The problem is that when the subscribers do not have a connection with the broker and messages are published, then the first one who connects receives not all messages, but only a part of them.

**Reply by obschepit** [reply]
Hi @zmstone !
have you reproduced the problem?

**Reply by zmstone** [reply]
Sorry. I haven’t managed to get time for this. I’ll ping my colleagues to see if someone else can help out.

**Reply by hjianbo** [reply]
Hi @obschepit I can't reproduce it too.

In step2, theoretically, emqx will still hold messages for ClientB on this topic after it disconnected. Instead, cast to ClientA.  After ClientB disconnected, did it reconnect before the publisher posted these messages?

BTW, is your emqx deployed as a single node or a cluster?

**Reply by obschepit** [reply]
Hi @HJianBo!

It's strange that you can't reproduce.

At step 2 emqx should not save messages for ClientB on this topic after it disconnected because it is a shared subscription and ClientA has a connection.
ClientB connected again only at step 3.4.
emqx deployed as a single node.

**Reply by hjianbo** [reply]
> At step 2 emqx should not save messages for ClientB on this topic after it disconnected because it is a shared 

However, in our implementation, if clientB's clean-session is false, the emqx will still deliver these messages to clientB's session as long as its session remains.

**Reply by hjianbo** [reply]
Can you open the emqx debug log? Then reproduce the process and send the log file to me??

**Reply by obschepit** [reply]
Hi.
here is my log file.
[erlang.log.1.zip](https://github.com/emqx/emqx/files/6365560/erlang.log.1.zip)

**Reply by obschepit** [reply]
hi.
guys, is there any news on this issue?
